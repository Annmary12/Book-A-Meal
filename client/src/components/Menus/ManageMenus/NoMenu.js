import React from 'react';
import menuImage from '../../../../assets/img/menu.svg';
import styles from './no-menu.scss';

const NoMenu = () => (
  <div className={styles.container}>
    <img
      src={menuImage}
      width="150"
      alt="no menu"
    />
    <p className={styles.text}>Hungry customers are waiting. Set a menu!</p>
  </div>
);

export default NoMenu;

