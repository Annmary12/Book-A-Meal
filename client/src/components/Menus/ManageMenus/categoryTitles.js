export default [
  'Appetizer',
  'Breakfast',
  'Lunch',
  'Dinner',
  'Dessert',
  'Drinks',
  'Specials'
];
