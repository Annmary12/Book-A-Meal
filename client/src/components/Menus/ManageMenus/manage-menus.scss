@import '../../../../assets/scss/_variables';
@import '../../../../assets/scss/buttons';

.container{
  margin: auto;
  width: 80%;
  @media only screen and (max-width: 600px) {
    margin-top: 86px;
  }
}

.menus {
  margin-top: 20px;
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(300px, 1fr));
}