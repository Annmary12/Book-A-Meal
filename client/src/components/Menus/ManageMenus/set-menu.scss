@import '../../../../assets/scss/_variables';
@import '../../../../assets/scss/form';
@import '../../../../assets/scss/buttons';

.meals {
  border: 2px solid #c0b8b8;
  border-radius: 5px;
  padding: 10px;
  padding-bottom: 1px;
  max-height: 95px;
  overflow: auto;
  position: relative;
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(200px, 1fr));
}

.successBtn {
  margin-right: 10px;
}

.searchInput{
  @extend .input;
  margin-bottom: 0px;
  border-bottom-left-radius: 0px;
  border-bottom-right-radius: 0px;
  border-bottom: 0px;
}

.menuCategory {
  border: 2px solid #c0b8b8;
  border-radius: 5px;
  padding: 5px;
  margin-bottom: 15px;
  box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24);

  :global{
    animation: fadeInUp 0.25s linear;
  }
}

.title {
  display: flex;
}

.right {
  margin-left: auto;
}

.modal{
  max-height: 668px;
  top: 41px;
}

.mealResult {
  background-color: white;
  z-index: 1;
  position: absolute;
  width: 100%;
  box-sizing: border-box;
  min-height: 95px;
  border: 2px solid #c0b8b8;
  border-bottom-left-radius: 5px;
  border-bottom-right-radius: 5px;
  max-height: 149px;
  overflow: auto;
  box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24);

  :global {
    animation: fadeInDown .07s ease-in;
  }

  ul {
    padding-left: 11px;
    list-style: none;
  }
}

.noMeal {
  text-align: center;
}

.error {
  border: 2px solid #de1e1e;
}

.mealsErrorMessage {
  color: #de1e1e;
  font-size: 13px;
  margin-bottom: 10px;
}