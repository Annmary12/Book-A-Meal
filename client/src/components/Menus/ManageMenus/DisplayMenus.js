import React from 'react';
import PropTypes from 'prop-types';
import MenuCardContainer from './MenuCard/MenuCardContainer';
import NoMenu from './NoMenu';
import styles from './manage-menus.scss';

const DisplayMenus = ({ menus }) => (
  <div className={styles.menus}>
    {
      menus.length ?
        menus.map(menu => <MenuCardContainer key={menu.id} menu={menu} />) :
        <NoMenu />
    }
  </div>
);

DisplayMenus.propTypes = {
  menus: PropTypes.arrayOf(PropTypes.object).isRequired
};

export default DisplayMenus;
