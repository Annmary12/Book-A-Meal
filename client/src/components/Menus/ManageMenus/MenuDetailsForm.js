import React, { Component } from 'react';
import PropTypes from 'prop-types';
import format from 'date-fns/format';
import { Link } from 'react-router-dom';
import categoryTitles from './categoryTitles.js';
import SetMenuCategory from './SetMenuCategory/SetMenuCategory.js';
import menuInfoSchema from '../../../utils/validation-schemas/menuInfoSchema.js';
import styles from './set-menu.scss';

/**
 * @class MenuDetailsForm
 */
class MenuDetailsForm extends Component {
  static propTypes = {
    closeModal: PropTypes.func.isRequired,
    handleSave: PropTypes.func.isRequired,
    meals: PropTypes.arrayOf(PropTypes.object).isRequired,
    date: PropTypes.string,
    categories: PropTypes.arrayOf(PropTypes.object)
  }

  static defaultProps = {
    date: '',
    categories: [{
      title: '',
      meals: []
    }]
  }

  state = {
    date: this.props.date,
    categories: this.props.categories,
    errors: {}
  }

  addCategory = () => {
    this.setState(state => ({
      categories: [
        ...state.categories,
        {
          title: '',
          meals: []
        }
      ]
    }));
  }

  handleDateChange = (event) => {
    this.setState({
      date: event.target.value
    });
  }

  handleCategoryTitleChange = index => (event) => {
    const { value } = event.target;

    this.setState((state) => {
      const categories = state.categories.slice();
      categories[index].title = value;

      return {
        categories
      };
    });
  }

  toggleMealInCategory = index => (meal) => {
    this.setState((state) => {
      const categories = state.categories.slice();
      let { meals } = categories[index];

      if (meals.includes(meal)) {
        meals = meals.filter(m => m.id !== meal.id);
      } else {
        meals.push(meal);
      }

      categories[index].meals = meals;
      return {
        categories
      };
    });
  }

  removeCategory = categoryIndex => () => {
    this.setState(state => ({
      categories: state.categories
        .filter((_, index) => index !== categoryIndex)
    }));
  }

  handleSave = () => {
    const { date, categories } = this.state;

    menuInfoSchema.validate({ date, categories }, { abortEarly: false })
      .then(() => {
        const modifiedCategories = categories.map((category) => {
          const modifiedCategory = {
            title: category.title,
            meals: category.meals.map(meal => meal.id).filter(Boolean)
          };

          return modifiedCategory;
        });

        return this.props.handleSave({ date, categories: modifiedCategories });
      })
      .then(this.props.closeModal)
      .catch((error) => {
        const errors = {};
        error.inner.forEach((err) => {
          errors[err.path] = err.message;
        });
        this.setState({
          errors
        });
      });
  }

  /**
   * @returns {JSX} React JSX
   * @memberof MenuDetailsForm
   */
  render() {
    const { closeModal, meals, date } = this.props;
    const { errors } = this.state;
    const selectedTitles = this.state.categories.map(c => c.title);

    return (
      meals.length ?
        <div>
          <h2>Menu details</h2>
          {
            date ?
              <h4>{format(date, 'Do MMM, YYYY')}</h4> :
              <label htmlFor="date" className={styles.label}>
                Date:
                <input
                  className={
                    errors.date
                      ? styles.inputError : styles.input}
                  name="date"
                  type="date"
                  min={format(new Date(), 'YYYY-MM-DD')}
                  value={this.state.date}
                  onChange={this.handleDateChange}
                />
                {
                  errors.date &&
                  <div className={styles.errorMessage}>{errors.date}</div>
                }
              </label>
          }

          {
          this.state.categories.map((category, index) => (
            <SetMenuCategory
              removeCategory={this.removeCategory(index)}
              handleCategoryTitleChange={this.handleCategoryTitleChange(index)}
              toggleMealInCategory={this.toggleMealInCategory(index)}
              category={category}
              selectedTitles={selectedTitles}
              categoryCount={this.state.categories.length}
              meals={meals}
              errors={errors}
              categoryIndex={index}
              key={index.toString()}
            />
          ))
        }

          <div className={styles.btns}>
            <div>
              <button
                className={styles.successBtn}
                onClick={this.handleSave}
              >
              Save
              </button>
              {
                this.state.categories.length < categoryTitles.length &&
                <button className={styles.closeModalBtn} onClick={this.addCategory}>
                  Add category
                </button>
              }
            </div>
            <button
              className={styles.closeModalBtn}
              onClick={closeModal}
            >
            Cancel
            </button>
          </div>
        </div> :
        <div className={styles.noMeal}>
          <p>You have not added a meal yet. Please add a meal to create a menu</p>
          <Link to="/manage-meals">Go to meals</Link>
        </div>

    );
  }
}

export default MenuDetailsForm;
