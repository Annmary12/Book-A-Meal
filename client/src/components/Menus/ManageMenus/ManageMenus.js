import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Loader from '../../Loader.js';
import styles from './manage-menus.scss';
import SetMenuModalContainer from './SetMenuModalContainer.js';
import DisplayMenus from './DisplayMenus.js';

/**
 * @class ManageMenus
 */
class ManageMenus extends Component {
  static propTypes = {
    getMenus: PropTypes.func.isRequired,
    getMeals: PropTypes.func.isRequired,
    menus: PropTypes.arrayOf(PropTypes.object).isRequired,
    isFetching: PropTypes.bool.isRequired,
    pagination: PropTypes.objectOf(PropTypes.number).isRequired
  }

  state = {
    setMenuModalOpen: false
  }

  /**
   * @returns {undefined} undefined
   */
  componentDidMount() {
    this.props.getMenus();
    this.props.getMeals();
  }

  closeMenuModal = () => {
    this.setState({ setMenuModalOpen: false });
  }

  openMenuModal = () => {
    this.setState({ setMenuModalOpen: true });
  }

  /**
   * @returns {JSX} React JSX
   */
  render() {
    return (
      <div className={styles.container}>
        <h2>Manage Menus</h2>
        <button
          className={styles.successBtn}
          onClick={this.openMenuModal}
        >
             Set Menu
        </button>
        {
              this.props.isFetching && this.props.menus.length < 1 ?
                <Loader /> :
                <DisplayMenus
                  menus={this.props.menus}
                  pagination={this.props.pagination}
                  getMenus={this.props.getMenus}
                />
            }
        <SetMenuModalContainer
          isOpen={this.state.setMenuModalOpen}
          handleClose={this.closeMenuModal}
        />
      </div>
    );
  }
}

export default ManageMenus;
