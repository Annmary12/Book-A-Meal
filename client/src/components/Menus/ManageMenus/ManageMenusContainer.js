import { connect } from 'react-redux';
import { getMenus } from '../../../actions/catererMenus.js';
import { getMeals } from '../../../actions/meals.js';
import ManageMenus from './ManageMenus.js';

const mapStateToProps = state => ({
  menus: state.catererMenus.menus,
  pagination: state.catererMenus.pagination,
  isFetching: state.catererMenus.isFetching
});

export default connect(mapStateToProps, { getMenus, getMeals })(ManageMenus);
