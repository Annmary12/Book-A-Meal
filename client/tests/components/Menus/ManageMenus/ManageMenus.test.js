import React from 'react';
import ManageMenus from '../../../../src/components/Menus/ManageMenus/ManageMenus';

const props = {
  getMenus: jest.fn(),
  getMeals: jest.fn(),
  menus: [],
  isFetching: true,
  pagination: {
    currentPage: 1
  }
}

describe('ManageMenus component', () => {
  it('should match snapshot if isFetching is true or menus array is empty', () => {
    const wrapper = shallow(
      <ManageMenus {...props}/>
    );

    expect(wrapper).toMatchSnapshot();
  });
  it('should match snapshot if isFetching is true and menus array is not empty', () => {
    const wrapper = shallow(
      <ManageMenus
        {...props}
        isFetching={false}
        menus={[{}]}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should call getMenus and getMeals prop function on mount', () => {
    const wrapper = shallow(
      <ManageMenus {...props}/>
    );

    expect(props.getMeals).toBeCalled();
    expect(props.getMenus).toBeCalled();
  });

  describe('closeMenuModal method', () => {
    const wrapper = shallow(
      <ManageMenus {...props}/>
    );

    it('should change setMenuModalOpen to false', () => {
      wrapper.instance().closeMenuModal();
      const { setMenuModalOpen } = wrapper.instance().state;

      expect(setMenuModalOpen).toBeFalsy();
    })
  })
  describe('openMenuModal method', () => {
    const wrapper = shallow(
      <ManageMenus {...props}/>
    );

    it('should change setMenuModalOpen to true', () => {
      wrapper.instance().openMenuModal();
      const { setMenuModalOpen } = wrapper.instance().state;

      expect(setMenuModalOpen).toBeTruthy();
    })
  })
});