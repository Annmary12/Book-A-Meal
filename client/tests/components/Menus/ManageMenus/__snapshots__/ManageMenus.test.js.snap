// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageMenus component should match snapshot if isFetching is true and menus array is not empty 1`] = `
<div
  className="container"
>
  <h2>
    Manage Menus
  </h2>
  <button
    className="successBtn"
    onClick={[Function]}
  >
    Set Menu
  </button>
  <DisplayMenus
    getMenus={
      [MockFunction] {
        "calls": Array [
          Array [],
          Array [],
          Array [],
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    menus={
      Array [
        Object {},
      ]
    }
    pagination={
      Object {
        "currentPage": 1,
      }
    }
  />
  <Connect(SetMenuModal)
    handleClose={[Function]}
    isOpen={false}
  />
</div>
`;

exports[`ManageMenus component should match snapshot if isFetching is true or menus array is empty 1`] = `
<div
  className="container"
>
  <h2>
    Manage Menus
  </h2>
  <button
    className="successBtn"
    onClick={[Function]}
  >
    Set Menu
  </button>
  <Loader
    message="loading..."
    width={40}
  />
  <Connect(SetMenuModal)
    handleClose={[Function]}
    isOpen={false}
  />
</div>
`;
